---
title: "BRAKER"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
    toc_float: true
    toc_collapsed: false
    theme: default
    css: "theme3.css"
toc_depth: 1
number_sections: true
editor_options:
  chunk_output_type: console
---

## Fully automated prediction of protein coding gene structure {.tabset}

### What is BRAKER?

From the Gaius-Augustus GitHub:

<b><i>"The rapidly growing number of sequenced genomes requires fully automated methods for accurate gene structure annotation. With this goal in mind, we have developed BRAKER1R1R0, a combination of GeneMark-ET R2 and AUGUSTUS R3, R4, that uses genomic and RNA-Seq data to automatically generate full gene structure annotations in novel genome.

However, the quality of RNA-Seq data that is available for annotating a novel genome is variable, and in some cases, RNA-Seq data is not available, at all.

BRAKER2 is an extension of BRAKER1 which allows for fully automated training of the gene prediction tools GeneMark-ES/ET/EP/ETP R14, R15, R17, F1 and AUGUSTUS from RNA-Seq and/or protein homology information, and that integrates the extrinsic evidence from RNA-Seq and protein homology information into the prediction.

In contrast to other available methods that rely on protein homology information, BRAKER2 reaches high gene prediction accuracy even in the absence of the annotation of very closely related species and in the absence of RNA-Seq data.

BRAKER3 is the latest pipeline in the BRAKER suite. It enables the usage of RNA-seq and protein data in a fully automated pipeline to train and predict highly reliable genes with GeneMark-ETP and AUGUSTUS. The result of the pipeline is the combined gene set of both gene prediction tools, which only contains genes with very high support from extrinsic evidence."</b></i>


### Keys to sucessful gene prediction

  - Use a high quality genome assembly. If you have a huge number of very short scaffolds in your genome assembly, those short scaffolds will likely increase runtime dramatically but will not increase prediction accuracy.
  
  - Use simple scaffold names in the genome file (e.g. *>contig1* will work better than *>contig1my custom species namesome putative function /more/information/  and lots of special characters %&!(){}*). Make the scaffold names in all your fasta files simple before running any alignment program.
  
  - In order to predict genes accurately in a novel genome, the genome should be masked for repeats. This will avoid the prediction of false positive gene structures in repetitive and low complexitiy regions. Repeat masking is also essential for mapping RNA-Seq data to a genome with some tools (other RNA-Seq mappers, such as HISAT2, ignore masking information). In case of GeneMark-ES/ET/EP/ETP and AUGUSTUS, softmasking (i.e. putting repeat regions into lower case letters and all other regions into upper case letters) leads to better results than hardmasking (i.e. replacing letters in repetitive regions by the letter N for unknown nucleotide).
  
  - Many genomes have gene structures that will be predicted accurately with standard parameters of GeneMark-ES/ET/EP/ETP and AUGUSTUS within BRAKER. However, some genomes have clade-specific features, i.e. special branch point model in fungi, or non-standard splice-site patterns. Please read the options section [options] in order to determine whether any of the custom options may improve gene prediction accuracy in the genome of your target species.
  
  - Always check gene prediction results before further usage! You can e.g. use a genome browser for visual inspection of gene models in context with extrinsic evidence data. BRAKER supports the generation of track data hubs for the UCSC Genome Browser with MakeHub for this purpose.
  
  
### Installation & Running

BRAKER is available on the Grace cluster, you can load it using the command 
```{r, eval=FALSE}
ml GCC/9.3.0  OpenMPI/4.0.3  BRAKER/3.0.3-Python-3.8.2

```

Then, there are many ways you can run BRAKER, but for our purposes we want to run with long-read RNA-Seq data and protein data. To do this we need to use the following steps that are based on the poster from Dr. Hoff [here](https://github.com/Gaius-Augustus/BRAKER/blob/master/docs/posters/poster_PAG2024.pdf):

  **1.** Download the iso-seq data by creating a file called *download-data.sh* and copy and paste the following code into it:
  ```{r, eval=FALSE}
#!/bin/bash

##NECESSARY JOB SPECIFICATIONS
#SBATCH --job-name=prefetch1          #Set the job name to "prefetch1"
#SBATCH --time=2-00:00:00             #Set the wall clock limit to 2 days
#SBATCH --ntasks=2                    #Request 2 task
#SBATCH --cpus-per-task=12            #Request 12 cpus for each task
#SBATCH --mem=50G                     #Request 50GB per node

# Load necessary modules
ml GCC/11.3.0  OpenMPI/4.1.4 SRA-Toolkit/3.0.3

# enable proxy to allow compute node connection to internet
module load WebProxy

# Define variables
input_file="SRR_Acc_List_isoseq_2024.txt"
output_dir="${Directory to BRAKER files}"

# Prefetch the data
prefetch --option-file "$input_file"

# Ensure output directory exists
mkdir -p "$output_dir"

# Read accession numbers and process with fasterq-dump and gzip
while IFS= read -r accession; do
  fastq-dump --split-3 --outdir "$output_dir" "$accession"
  if [[ $? -ne 0 ]]; then
    echo "Error: Processing accession $accession encountered an error."
    exit 1
  fi

  # Compress the resulting fastq files
  for file in "$output_dir"/*; do
    if [[ -f $file && $file == *.fastq ]]; then
      gzip "$file"
      if [[ $? -ne 0 ]]; then
        echo "Error: Compressing file $file encountered an error."
        exit 1
      fi
    fi
  done
done < "$input_file"

echo "fasterq-dump and gzip completed successfully."
  
  ```
  
  Then visit [this site](https://www.ncbi.nlm.nih.gov/sra/) to get the IsoSeq data SRR access numbers. Select all available files by checking the box next to all the files and selecting Send > File > Accession List to in the upper right hand corner. This will download a .csv file with all of the accession numbers for your data. You then need to create a file called *SRR_Acc_List_isoseq.txt* and copy and paste the accession numbers into it. You can then run the *download-data.sh* file.
  
  <br>
  **2.** Now you must combine all of the IsoSeq data, map it to your genome of choice, and then convert the sam file to a BAM file by copy and pasting the following script into a file called *run-alignment.sh*:
  ```{r, eval=FALSE}
  
  #!/bin/bash
  #SBATCH --job-name=braker-step2           # Job name
  #SBATCH --output=braker-step2.out         # Output file
  #SBATCH --error=braker-step2.err          # Error file
  #SBATCH --nodes=1                         # Number of nodes
  #SBATCH --ntasks=1                        # Number of tasks
  #SBATCH --cpus-per-task=12                # Number of CPU cores per task
  #SBATCH --mem=200G                        # Total memory limit
  #SBATCH --time=48:00:00                   # Time limit hrs:min:sec

  # Load necessary modules
  module load GCC/13.2.0 GCCcore/12.2.0 minimap2/2.26 SAMtools/13.2.0
  
  # Concatenate FASTQ files
  cat file-1.fastq file-2.fastq ... file-n.fastq > all-subreads.fa
  
  # Run minimap2 for alignment w/ increased memory allotment
  minimap2 -t 12 -ax splice:hq -I4Gb --split-prefix=tmp GENOME.fna all-subreads.fa > long_reads.sam
  
  # Use Samtools to convert sam to BAM
  samtools view -bS --threads 48 long_reads.sam -o isoseq.bam

  ```
  <br>
  **3.** Build the Singularity image
  ```{r, eval=FALSE}
  singularity build braker3_lr.sif docker://teambraker/braker3:isoseq
  ```  
  <br>
  **4.** Call BRAKER3 with the BAM file
  ```{r, eval=FALSE}
  singularity exec -B ${PWD}:${PWD} braker3_lr.sif braker.pl --genome=genome.fa --prot_seq=protein_db.fa --bam=isoseq.bam --threads=48
  ```
  
### Troubleshooting

#### Samtools: [E::sam_parse1] no SQ lines present in the header
SQ lines are used by Samtools to align sequences in .sam and BAM files. This error occurs when SQ lines are not included in the .sam file and the you try to convert it to a BAM file. To solve this issue, add the flag --split-prefix=tmp to split up the alignments in order to retain these SQ lines in the headers (I have now included this in my Step 2 command by default).

### Citations 
Stanke, M., Diekhans, M., Baertsch, R. and Haussler, D. (2008). *Using native and syntenically mapped cDNA alignments to improve de novo gene finding*. Bioinformatics, doi: 10.1093/bioinformatics/btn013.

Stanke. M., Sch√∂ffmann, O., Morgenstern, B. and Waack, S. (2006). *Gene prediction in eukaryotes with a generalized hidden Markov model that uses hints from external sources*. BMC Bioinformatics 7, 62.
