---
title: "Compare_Genomes"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
    toc_float: true
    toc_collapsed: false
    theme: default
    css: "theme3.css"
toc_depth: 1
number_sections: true
editor_options:
  chunk_output_type: console
---

## A Comparative Genomics Workflow to Streamline the Analysis of Evolutionary Divergence Across Eukaryotic Genomes {.tabset}
### What is Compare_Genomes?
>The dawn of cost-effective genome assembly is enabling deep comparative genomics to address fundamental evolutionary questions by comparing the genomes of multiple species. However, comparative genomics analyses frequently deploy multiple, often purpose-built frameworks, limiting their transferability and replicability. Here, we present compare_genomes, a transferable and extensible comparative genomics workflow package we developed that streamlines the identification of orthologous families within and across eukaryotic genomes and tests for the presence of several mechanisms of evolution (gene family expansion or contraction and substitution rates within protein-coding sequences). The workflow is available for Linux, written as a Nextflow workflow that calls established genomics and phylogenetics tools to streamline the analysis and visualization of eukaryotic genome divergence.

--- Paril, J., Zare, T., & Fournier‐Level, A. (2023).*Compare_genomes: A comparative genomics workflow to streamline the analysis of evolutionary divergence across eukaryotic genomes*

#### My thoughts on the pipeline
So far the instructions are incredibly clear and have made it a lot easier than a lot of the other pipelines to install and get running. I love that it comes with a .yml file that creates and configures the conda environment for you, that alone gets rid of a lot of sources of error when it comes to getting it properly configured to run. Paril even includes the commands you need to edit the run.sh file directly from the terminal. So far, this has been easier than any other pipeline I've tried to use before. The only grievance I have is that it requires a lot of downloads to run, but as long as it's not having to be done repeatedly it should be okay.

#### Workflow of the pipeline
This is taken directly from the paper by Paril, Zare, and Fournier-Level:

"The compare_genomes workflow consists of nine analysis steps under the default setup (Fig. 1, left).

a. Download the user-defined genome datasets: genome sequences (fasta, .fna), annotations (general feature format, .gff), coding DNA sequences or CDS (.cds), protein sequences (fasta, .faa), protein-coding gene models (probabilistic protein model format, .hmm), corresponding gene ontology terms (.txt), and protein sequences of specific genes of interest (.faa).

b. Identify orthogroups using OrthoFinder (Emms & Kelly, 2019) and gene families for each orthogroup using HMMER3 (Mistry et al., 2013) and Panther HMMs (protein-coding gene family models; Thomas et al., 2022). An orthogroup is a set of genes descended from a single gene from the last common ancestor of all the species included in the analysis.

c. Infer phylogenetic trees for each orthogroup using IQ-TREE 2 (Minh et al., 2020) based on CDS alignments generated by MACSE (Ranwez et al., 2011) and the most likely nucleotide substitution model inferred by ModelFinder (Kalyaanamoorthy et al., 2017).

d. Infer the rate of sequence divergence based on transversion rates among four-fold-degenerate sites (4DTv) in single-copy genes between pairs of species using the custom Julia script for this purpose that ships with compare_genomes. 4DTv is a proxy for time, where the accumulation of transversion mutations at the degenerate (neutral) site is proportional to the amount of time passed.

e. Identify whole-genome duplication events using 4DTv computed from multi-copy gene families. If the paralogs within a genome show an accumulation of 4DTv that is greater than 0, it is likely that a genome duplication or polyploidization occurred from which the multiple paralogs are derived.

f. Test for significant gene family expansion or contraction across genomes using CAFE (version 5; De Bie et al., 2006). Gene family expansion and contraction are calculated relative to the gene family count in the ancestral species given the phylogenetic tree from IQ-TREE 2. Expanded and contracted gene families can be indicative of adaptation.

g. Analyze gene ontology (GO) term enrichment for significantly expanded gene families using the Panther GO API (Mi et al., 2019). To test the significantly contracted gene families, replace all instances of the term ‘expanded’ with ‘contracted’ in lines 47-59 of ‘compare_genomes/modules/GO_enrichment.nf’.

h. Visualize a summary of the results (i.e., see Fig. 1, right, for a sample output). This generates the summary output of the whole-genome-level analysis.

i. This optional step is available for testing hypotheses involving specific genes: analyze user-defined genes of interest, i.e., gene family expansion/contraction analyses with CAFE, and estimate non-synonymous to synonymous nucleotide substitution rates (Ka/Ks) using KaKs_Calculator 2.0 (Wang et al., 2009) and custom R script."

### Installation and Running
This pipeline is not available on Grace, but it is pretty simple to create a conda environment and run it from there by using these steps. The only difference between running locally and on a cluster will occur in step 2.

**1.** Download the compare_genomes repository using the command
```{r repos-dnld, eval=FALSE}
git clone https://github.com/jeffersonfparil/compare_genomes.git

```

**2.** 

<span style="text-decoration:underline">Locally:</span> Install Conda (<mark>Only if you have not installed it yet. If you have, you can skip this step.</mark>)
```{r install-conda, eval=FALSE}
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
sh./Miniconda3-latest-Linux-x86_64.sh
```

<span style="text-decoration:underline">On a slurm cluster:</span> Load the Anaconda module
```{r act-conda, eval=FALSE}
ml Anaconda3/2022.10 # Or whatever recent version is available 
```

**3.** Import and activate the compare_genomes Conda environment:
```{r imprt-env, eval=FALSE}
conda env create -n compare_genomes --file compare_genomes/compare_genomes.yml

conda activate compare_genomes
```
This step may take a bit as there are a lot of modules that need to be downloaded.

**4.** Edit the configuration file, i.e., compare_genomes/config/params.config, by replacing line number 2, dir = ‘/data/TEST’, with the absolute path on the user's computer that will be used as the output directory for the workflow:

```{r edit-config, eval=FALSE}
cd compare_genomes
nano config/params.config
Replace “dir='/data/TEST'” with your path.
```
Possible source of error: <mark> Replace /data/TEST with the *entire* directory path of your chosen output file. 

**5.** Run the example:
```{r run-ex, eval=FALSE}
chmod +x run.sh
time ./run.sh
```


#### Setting up for a new analysis
Paril considers this step 6 of running the pipeline, so if you read the [paper](https://currentprotocols.onlinelibrary.wiley.com/doi/full/10.1002/cpz1.876) that is where you will find these steps. The steps below are taken directly from their paper.

To set up the workflow for a new analysis, edit the seven configuration files located in compare_genomes/config/.

a. ‘urls.txt’: list of web links or absolute paths to the genome sequences, genome annotations, coding DNA sequences, and amino acid sequences for at least three species to be included in the analyses. It is formatted as a headerless, two-columned, comma-separated file. Column 1 contains the filenames of the genome sequence, genome annotation, coding DNA sequence, and amino acid sequences (species names and extension names should be the consistent across these files, e.g., ‘.fna’ for the genomes, ‘.gff’ for the annotations, ‘.cds’ for the coding DNA sequences, and ‘.faa’ for the amino acid sequences). Column 2 contains the URL (uniform resource locator) of the zipped (‘.gz’ or ‘.zip’) or unzipped files for download. Alternatively, this can be the absolute path to the pre-downloaded zipped or unzipped files on a local computer.

b. ‘dates.txt’: list of pairwise divergence times between species. This information can be found at http://timetree.org. Divergence times between all pairs of species are not required. Pick at least two pairs of species divergence times, ideally including the outgroup species. This file is formatted as a headerless, two-columned, tab-delimited file. Column 1 contains the pair of species separated by a comma with the same names used in the ‘urls.txt’. Column 2 contains the time in million years, e.g., ‘-160’ for 160 million years ago.

c. ‘comparisons_4DTv.txt’: list of species and pairs of species to be included in the estimation of transversion rates among four-fold-degenerate sites (4DTv). This statistic is used to set the molecular clock, with more mutations at the third codon position meaning more divergence time between a pair of sequences. By default, 4DTv is estimated using genes present with two copies. Edit line 34 of ‘compare_genomes/modules/assess_WGD.nf’ to include genes with more than two copies. This is formatted as a headerless, one-columned file. Column 1 contains the species and/or pairs of species names, which should match the names in ‘urls.txt’, and species pairs should be written as, for example, “Zea_mays X Oryza_sativa”.

d. ‘venn_species_max_5.txt’: list of at most five species to be plotted in the Venn diagram comparing the differences and commonalities of gene families between species. It is currently not possible to fit more than five species in the Venn diagram because of the limitations of the plotting package used. This is formatted as a headerless, one-columned file. Column 1 contains the species names matching those in ‘urls.txt’.

e. ‘genes.txt’: links to the gene sequences to be tested for significant expansion/contraction and for nonsynonymous/synonymous mutation (Ka/Ks) rates between pairs of sequences within and among species. It is formatted as a headerless, three-columned, comma-separated file. Column 1 contains phenotype names or some identifier (noncritical information). Column 2 contains the species names from which the gene sequence was derived and can be a species not included in ‘urls.txt’ (noncritical information). Column 3 contains the URL of the genes to be downloaded and analyzed.

f. ‘params.config’: configuration file listing the parameter values for the specific analyses to be conducted.

    i. ‘dir’: output directory.
    ii. ‘species_of_interest’: a single species of interest, which should match one of the species listed in ‘urls.txt’.
    iii. ‘species_of_interest_panther_HMM_for_gene_names_url’: URL to the Panther HMM database to extract gene names from, preferably from the species used for the gene ontology (GO) term enrichment analysis. See the current release list at http://data.pantherdb.org/ftp/sequence_classifications/current_release/PANTHER_Sequence_Classification_files/.
    iv. ‘urls’: location of ‘urls.txt’.
    v. ‘dates’: location of ‘dates.txt’.
    vi. ‘comparisons_4DTv’: location of ‘comparisons_4DTv.txt’.
    vii. ‘venn_species_max_5’: location of ‘venn_species_max_5.txt’.
    viii. ‘genes’: location of ‘genes.txt’.
    ix. ‘cafe5_n_gamma_cats’: number of the Gamma values (parameter of the substitution model) to use for the assessment of significant gene family expansion and contraction using CAFE5. If this is equal to 1, then we use the substitution model without the gamma function.
    x. ‘cafe5_pvalue’: significance threshold for the gene family expansion and contraction test.
    xi. ‘go_term_enrich_genome_id’: genome ID for the species specified in ‘species_of_interest_panther_HMM_for_gene_names_url’ or some closely related species. Find the appropriate taxon ID at https://pantherdb.org/services/oai/pantherdb/supportedgenomes.
    xii. ‘go_term_enrich_annotation_id’: code for the gene ontology level to be used, e.g., “GO:0008150” for "Biological Process". See the list of GO codes at https://pantherdb.org/services/oai/pantherdb/supportedannotdatasets.
    xiii. ‘go_term_enrich_test’: GO term enrichment test, which can be set to either “FISHER” (Fisher's exact test) or “BINOMIAL” (binomial distribution test).
    xiv. ‘go_term_enrich_correction’: multiple testing correction which can be set to “NONE”, “FDR” (false-discovery rate), or “BONFERRONI” (Bonferroni correction).
    xv. ‘go_term_enrich_ngenes_per_test’: number of randomly sampled genes to include in each GO term enrichment analysis.
    xvi. ‘go_term_enrich_ntests’: number of GO term enrichment test replications to perform.
g. ‘process.config’: configuration file setting the computing resource allocation. Assign the number of ‘cpus’ and ‘memory’ capacity to use for low- and high-resource tasks with ‘LOW_MEM_LOW_CPU’ and ‘HIGH_MEM_HIGH_CPU’, respectively.

### Troubleshooting

####  Pipeline cannot find output directory you specified in the config file
Make sure you put the *entire* directory path in the params.config file. Yes, even if you're in a "closer" directory, you have to put the full file path from the very first parent folder, down to the output folder you've created. For example, I would enter the *entire* file path shown in the light blue box below if I wanted compare_genomes to output to my folder called output. 

```{r, echo=FALSE}
knitr::include_graphics("analysis/file-path.png")
```




