---
title: "make_lastz_chains"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
    toc_float: true
    toc_collapsed: false
    theme: default
    css: "theme3.css"
toc_depth: 1
number_sections: true
editor_options:
  chunk_output_type: console
---

## Creating Genome Alignments for TOGA {.tabset}

### What is make_lastz_chains?

>This program is also from Hiller Lab (see [TOGA](toga.html)) and it is a pipeline that takes in a reference and a query genome either as .fna or .2bit files and creates pairwise genome alignment pairs from them for use in TOGA. 

Read more in depth on their GitHub [here](https://github.com/hillerlab/make_lastz_chains).

### Installation and Running
<center>
<mark style="background: #f9d9da">This program is supported on MacOS, but the Hiller Lab team *highly   
recommends* running it on a Linux-based machine.</mark>
</center>

#### Prepare Files

To acquire files to put into this pipeline, find your genome(s) of interest on NCBI, click on the FTP tab, and download the file ending in genomic.fna.gz. Then create a folder called *data* and copy your files from your system into this folder. Then run the command

```{r gunzip, eval=FALSE}

gunzip YOUR-GENOME_genomic.fna.gz

```

This will turn your file from a .fna.gz to just .fna. Then create a file called *remove-headers.sh* in your *data* folder and copy the following code into it.

```{r remove-headers.sh, eval=FALSE}
#!/bin/bash

# Define the input and output files
input_file="input-name.fna"
output_file="input-name_only_contig_name.fna"

# Use sed to process both headers and sequences in one pass
sed -e 's/^\(>\S*\).*/\1/' "$input_file" > "$output_file"

echo "Processing complete. Output written to $output_file."

```

This script removes everything except the contig name from the headers in the .fna file. We do this because make_lastz_chains will through a spacing error otherwise. <u> You must do this for each file you intend to create chains for.</u> You are now ready to run this pipeline!

You can also run RepeatMasker and RepeatModeler on your genome files. The compatible versions are RepeatMasker/4.1.4 and RepeatModeler/2.0.5.

#### Running on Grace
To run the make_lastz_chains pipeline on Grace, create a file in your folder for make_lastz_chains called *run-mlc.sh* and copy and paste the following code into it:
```{r act-mlc, eval=FALSE}
#!/bin/sh
#SBATCH --time=0-1:00:00
#SBATCH --ntasks=48
#SBATCH --mem=50G
#SBATCH --nodes=1
#SBATCH --job-name=MLC-TEST

# module purge

module load GCC/12.3.0  make_lastz_chains/1.0.0  # load modules

conda activate mlc

./make_chains.py target query ${Path to Target Genome} ${Path to Query Genome} --pd test_out -f --chaining_memory 16 --executor slurm

```

Then run this script using the command
```{r run-mlc-command, eval=FALSE}
sbatch run-mlc.sh
```

#### Running in a Conda Environment
Follow the instructions on the Hiller Lab GitHub [here](https://github.com/hillerlab/make_lastz_chains) to install.

### Troubleshooting
####  fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists.

The full error is here:
```{r remote-fatal-fail, eval=FALSE}
Warning: Permanently added the ECDSA host key for IP address '140.82.113.3' to the list of known hosts. Permission denied (publickey). fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists.
```

I corrected this by going into my home > .ssh directory in Grace and adding the passkey from *id_rsa.pub* to my GitHub SSH keys. This is the public key the Grace terminal uses to download files from GitHub. This allowed me to download from the Hiller Lab repository.  

If your *id_rsa.pub* file is empty, visit [this link](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/Fix-GitHubs-Permission-denied-publickey-SSH-error). It has the full documentation that I followed to fix this error. Here you will find instructions on how to get the necessary information in your *id_rsa.pub* file.

#### Error: detected space-or-tab-containing sequence

The whole error is below:

```{r space-error, eval=FALSE}
Error! File: data/Nitens-and-Greg/GCF_021461395.2_iqSchAmer2.1_genomic.fna - detected space-or-tab-containing sequence:
NC_060119.1 Schistocerca americana isolate TAMUIC-IGC-003095 chromosome 1, iqSchAmer2.1, whole genome shotgun sequence
Please exclude or fix sequences with spaces and tabs.
```

To correct this error, first run your .fna files through this script:

```{r space-error-correcting-script, eval=FALSE}
#!/bin/bash

# Define the input and output files
input_file="input-name.fna"
output_file="input-name_only_contig_name.fna"

# Use sed to process both headers and sequences in one pass
sed -e 's/^\(>\S*\).*/\1/' "$input_file" > "$output_file"

echo "Processing complete. Output written to $output_file."
```

This removes everything following the contig name in each header of the fasta file. 

#### Error: unrecognized arguments: -f

The -f or -force flag is meant to overwrite an output folder if a folder exists with the same name already. This error occurs when there is no pre-existing folder to overwrite, so simply remove this flag from your command and rerun.


#### ModuleNotFoundError: No module named 'dataclasses'

This error occurs for me when the incorrect GCC module is loaded on Grace, verify that GCC 12.3.0 is available on the cluster and if not contact IT. 

If running locally, verify that the correct GCC is loaded on your machine.

#### Error! The nextflow process lastz crashed!
This error is addressed on the Hiller Lab GitHub page [here](https://github.com/hillerlab/make_lastz_chains/issues/16), but here is the comment from Dr. Hiller containing both the source of the issue and the solution: 
"I think, based on the log file, you don't get to the chaining step, as some of the lastz jobs fail (increasing chaining mem is downstream).
This typically happens when lastz 'sees' too many seeds from repeats.

If you already repeatModeled (a denovo RepeatModeler run is important) and masked your genome, you could try reducing chunksize (e.g. set
--seq1_chunk 50000000
--seq2_chunk 10000000
). Gives you more but smaller jobs."

So by adding the --seq1_chunk 50000000 and --seq2_chunk 10000000 flags to your *run-mlc.sh* file, this problem should be fixed.

#### Number of jobs should not exceed 6000
Run the command
```{r correct-numbr-of-jobs, eval=FALSE}
rm -rf /scratch/group/songlab/emily/comparative-genomics/make_lastz_chains/greg-and-serialis/TEMP_run.lastz /scratch/group/songlab/emily/comparative-genomics/make_lastz_chains/greg-and-serialis/TEMP_psl
```
This command is provided by the error message in the make_lastz_chains module. If this does not work, increase your node memory in your slurm script and the chaining memory in your ./run-mlc.sh command. For example, IT had me increase my node memory from 50G to 360G and my chaining memory from 16 to 200.





### Citations
Kirilenko BM, Munegowda C, Osipova E, Jebb D, Sharma V, Blumer M, Morales A, Ahmed AW, Kontopoulos DG, Hilgers L, Lindblad-Toh K, Karlsson EK, Zoonomia Consortium, Hiller M. *Integrating gene annotation with orthology inference at scale.*, Science, 380, 2023  

Osipova E, Hecker N, Hiller M. *RepeatFiller newly identifies megabases of aligning repetitive sequences and improves annotations of conserved non-exonic elements*. GigaScience, 8(11), giz132, 2019  
  
Suarez H, Langer BE, Ladde P, Hiller M. *chainCleaner improves genome alignment specificity and sensitivity*. Bioinformatics, 33(11), 1596-1603, 2017  

Kent WJ, Baertsch R, Hinrichs A, Miller W, Haussler D. *Evolution's cauldron: Duplication, deletion, and rearrangement in the mouse and human genomes*, PNAS, 100(20):11484-9, 2003
