---
title: "TOGA: Tool to Infer Orthologs from Genome Alignments"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: false
    toc_float: true
    toc_collapsed: false
    theme: default
    css: "theme3.css"
toc_depth: 1
number_sections: true
editor_options:
  chunk_output_type: console
---

## TOGA {.tabset}
### What is TOGA?



<div class="img-with-text">

<br>
<img src="assets/toga-logo.png" alt="TOGA Logo">
<div class="text">
<b>"Annotating coding genes and inferring orthologs are two classical challenges in genomics and evolutionary biology that have traditionally been approached separately, limiting scalability. We present TOGA (Tool to infer Orthologs from Genome Alignments), a method that integrates structural gene annotation and orthology inference. TOGA implements a different paradigm to infer orthologous loci, improves ortholog detection and annotation of conserved genes compared with state-of-the-art methods, and handles even highly fragmented assemblies."</b>

  -- Kirilenko <i>et al.</i>, <i>Integrating gene annotation with orthology inference at scale</i>
  
</div>


</div>
<style>
.img-with-text {
  display: flex;
  align-items: flex-start;
}

.img-with-text img {
  margin-right: 15px;
  max-width: 500px;
  height: auto;
}

.img-with-text .text {
  flex: 3 0 450px;
  margin-left: 15px;
  display: flex;
  flex-direction: column;
}
</style>

<hr class="solid">

###### Steps of TOGA

1. First, TOGA uses pairwise genome alignment between the reference & query genome represented by chains of co-linear local alignments. Through this process, it captures orthologous gene loci and loci containing paralogs or processed pseudo-genes. To distinguish between these 2 groups, TOGA computes characteristic features that capture the amount of intronic and intergenic alignments, considering each gene & overlapping chain. Additionally, synteny (conserved gene order) can be used to help further distinguish orthologs from paralogs. TOGA also implements machine learning to compute the probability that a chain is an orthologous locus for the gene in question. 

  - The most important feature for classification is the capture of    intronic and intergenic alignments, leaving synteny as the least    important. Hence, synteny is used as an auxiliary feature that      allows TOGA to detect orthologs that underwent translocation or     inversion.

  * <details>
     <summary>More on the machine learning feature</summary>
       
       The machine learning classifier was trained on known orthologous genes between humans (reference) and mice (query) from Ensembl Compara. This classifier was then tested on independent query species (rat, dog, and armadillo) that represented different placental mammal orders  and produced a nearly perfect classification of orthologous chains. 
       
       When manually examining data it was found that misclassifications that produced false positives are mostly due to partial or full gene duplications and half of false negative may be due to faster X chromosome evolution. 
       
   </details>
   
2. CESAR v. 2.0 is used to determine the position of coding exons of the focal gene in each (co-)orthologous query locus. TOGA then assesses the reading frame intactness for each transcript and applies a gene loss detection approach to identify gene-inactivating mutations while taking assembly incompleteness into account. A gene is classified as *lost* if all transcripts at all (co-)orthologous loci are classified as lost.

  * <details>
      <summary>More on CESAR v. 2.0</summary>
      "CESAR 2.0 (Coding Exon Structure Aware Realigner 2.0) is a          method to realign coding exons or genes to DNA sequences            using a Hidden Markov Model.  

      Compared to its predecessor, CESAR 2.0 is 77X times faster       on average (132X times faster for large exons) and requires         30-times less memory. In addition, CESAR 2.0 improves the           accuracy of the comparative gene annotation by two new              features. First, CESAR 2.0 substantially improves the               identification of splice sites that have shifted over a larger       distance, which improves the accuracy of detecting the correct       exon boundaries. Second, CESAR 2.0 provides a new gene mode         that re-aligns entire genes at once. This mode is able to           recognize complete intron deletions and will annotate larger        joined exons that arose by intron deletion events."
      
      -- From the [Hiller Lab GitHub page on CESAR 2.0](https://github.com/hillerlab/CESAR2.0)
      
    </details>
    
3. In this step, TOGA determines the orthology type by considering all reference genes and orthologous query loci that encode within an intact reading frame. TOGA then uses an orthology graph to resolve any weakly supported orthology relationship among many:many orthologs.

### Installation & Running

<mark>These instructions are specific to the Grace cluster at Texas A&M University, you may need to contact your IT for specifics regarding steps 2 and 3. Read more about the TAMU Grace cluster specs [here](https://hprc.tamu.edu/kb/User-Guides/Grace/) to compare to your own setup. </mark> 

<u>These instructions are courtesy of Michael Dickens at TAMU Grace, who was a *huge* help getting this pipeline running!</u>  

1. Clone the Hiller Lab GitHub into your own directory with the command

```{r clone-github, eval=FALSE}
mkdir $SCRATCH/toga_test
cd $SCRATCH/toga_test/
git clone https://github.com/hillerlab/TOGA.git
cd TOGA/
```

2. Edit file with the following 5 lines (change executor.queueSize to 100): nextflow_config_files/call_cesar_config_template.nf

```{r edit-cesar-config, eval=FALSE}
executor.queueSize = 100
process.pollInterval = '3 min'
process.queueStatInterval = '6 min'
process.submitRateLimit = '2/1min'
process.clusterOptions = '--ntasks=1 --cpus-per-task=1'

```

3. Edit file with the following 4 lines: nextflow_config_files/extract_chain_features_config.nf 

```{r edit-nextflow-config, eval=FALSE}
process.pollInterval = '3 min'
process.queueStatInterval = '6 min'
process.submitRateLimit = '2/1min'
process.clusterOptions = '--ntasks=1 --cpus-per-task=1'
```

4. Copy the 2bit files

```{r copy-2bits, eval=FALSE}
cp /scratch/data/bio/ucsc/gbdb/hg38/hg38.2bit  test_input/
cp /scratch/data/bio/ucsc/gbdb/mm10/mm10.2bit  test_input/ 
```

5. Load the TOGA module and run the configure.sh script

```{r load-mods-michael, eval=FALSE}
module purge
module load GCC/11.3.0  OpenMPI/4.1.4  TOGA/2024.05.26  git/2.36.0-nodocs
./configure.sh
```

6. Create a job script in the TOGA directory with the following lines and submit the job script using sbatch:
```{r michael-toga-script, eval=FALSE}
#!/bin/sh
#SBATCH --job-name=toga
#SBATCH --time=7-00:00:00
#SBATCH --nodes=1
#SBATCH --ntasks=8
#SBATCH --mem=50G

module purge
module load GCC/11.3.0  OpenMPI/4.1.4  TOGA/2024.05.26 

./toga.py $PWD/test_input/hg38.mm10.chr11.chain $PWD/test_input/hg38.genCode27.chr11.bed $PWD/test_input/hg38.2bit $PWD/test_input/mm10.2bit --kt --pn test -i $PWD/supply/hg38.wgEncodeGencodeCompV34.isoforms.txt --nc $PWD/nextflow_config_files --cb 3,5 --cjn 500 --u12 $PWD/supply/hg38.U12sites.tsv --ms
```

###### Testing

The Hiller Lab provides a test [reference](https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.2bit) and [query](https://hgdownload.cse.ucsc.edu/goldenpath/mm10/bigZips/mm10.2bit) genome, already in the required 2-bit formatting. You will also need to download the [supply](https://github.com/hillerlab/TOGA/tree/master/supply) folder in order to have the U12 data for the genomes. You can download the whole folder at once by copy and pasting the URL into [this website](https://download-directory.github.io/).

After this data has been downloaded & placed in the correct directory, call the following command:

```{r test, eval=FALSE}
./toga.py test_input/hg38.mm10.chr11.chain test_input/hg38.genCode27.chr11.bed ${path_to_human_2bit} ${path_to_mouse_2bit} --kt --pn test -i supply/hg38.wgEncodeGencodeCompV34.isoforms.txt --nc ${path_to_nextflow_config_dir} --cb 3,5 --cjn 500 --u12 supply/hg38.U12sites.tsv --ms

```

This will take about 20 minutes on a 500 core cluster.

### File Formatting
In order to format files correctly, you must 

1. Download the Hiller Lab's make_lastz_chain pipeline by following the instructions [here](https://github.com/hillerlab/make_lastz_chains). 

2. <u>If you are running TOGA on Grace, skip this step! </u> Install the  necessary utilities. Some of these are necessary to run the pipeline and some will be used to convert our fasta files to the correct formats. You can download all of them using this command:

```{r convert-downloads, eval=FALSE}
Mac download: rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/macOSX.x86_64/ ./
Windows or Linux download: rsync -aP rsync://hgdownload.soe.ucsc.edu/genome/admin/exe/linux.x86_64/ ./
```


### Troubleshooting

#### Grace: PermissionError: [Errno 13]
  
The full input and error message is here: 
  
```{r error, eval=FALSE}
  (base) [ecbaker7@c076 toga]$ toga.py test_input/hg38.mm10.chr11.chain test_input/hg38.genCode27.chr11.bed $/toga $/toga --kt --pn test -i supply/hg38.wgEncodeGencodeCompV34.isoforms.txt --nc $/scratch/group/songlab/emily/toga/nextflow_config_files --cb 3,5 --cjn 500 --u12 supply/hg38.U12sites.tsv --ms  
###### Initiating TOGA class ######
Checking dependencies...
check if binaries are compiled and libs are installed...
All dependencies found
/bin/sh: git: command not found
Traceback (most recent call last):
  File "/sw/eb/sw/TOGA/2021.12.15-foss-2019b-Python-3.7.4/toga.py", line 1965, in <module>
    main()
  File "/sw/eb/sw/TOGA/2021.12.15-foss-2019b-Python-3.7.4/toga.py", line 1960, in main
    toga_manager = Toga(args)
  File "/sw/eb/sw/TOGA/2021.12.15-foss-2019b-Python-3.7.4/toga.py", line 88, in __init__
    self.nextflow_dir = self.__get_nf_dir(args.nextflow_dir)
  File "/sw/eb/sw/TOGA/2021.12.15-foss-2019b-Python-3.7.4/toga.py", line 362, in __get_nf_dir
    os.mkdir(default_dir) if not os.path.isdir(default_dir) else None
PermissionError: [Errno 13] Permission denied: '/sw/eb/sw/TOGA/2021.12.15-foss-2019b-Python-3.7.4/nextflow_logs'
```


I corrected this by creating a new directory called *nextflow_logs* and then adding the flag --nextflow_dir or --nd followed by the path to *nextflow_logs*. This gives Nextflow a directory to operate from. The full command is:

```{r corrected-command, eval=FALSE}

toga.py test_input/hg38.mm10.chr11.chain test_input/hg38.genCode27.chr11.bed test_input/hg38.2bit test_input/mm10.2bit --kt --pn test -i supply/hg38.wgEncodeGencodeCompV34.isoforms.txt --nc nextflow_config_files --cb 3,5 --cjn 500 --u12 supply/hg38.U12sites.tsv --ms --nextflow_dir nextflow_logs

```

#### Error on Toga step 2: Nextflow jobs won't execute

I'm having an error where I get to step 2 of TOGA but it can't complete any jobs and keeps retrying all of them. The error message I get in my .out file is 

```{r nf-no-jobs, eval=FALSE}
Traceback (most recent call last):
  File "/var/spool/slurmd/job10522779/slurm_script", line 19, in <module>
    from modules.common import parts
ModuleNotFoundError: No module named 'modules'

``` 

I corrected this error by running the job through a .sh file called *run-toga.sh* using sbatch. It contained the commands below:

```{r run-toga, eval=FALSE}
#!/bin/sh
ml GCC/8.3.0 OpenMPI/3.1.4 TOGA/2021.12.15-Python-3.7.4		 # load all necessary modules

toga.py test_input/hg38.mm10.chr11.chain test_input/hg38.genCode27.chr11.bed test_input/hg38.2bit test_input/mm10.2bit --kt --pn test -i supply/hg38.wgEncodeGencodeCompV34.isoforms.txt --nc nextflow_config_files --cb 3,5 --cjn 500 --u12 supply/hg38.U12sites.tsv --ms --nextflow_dir nextflow_logs
```

You can then run *run-toga.sh* on the cluster using the command
```{r comm-run-toga, eval=FALSE}
sbatch --ntasks=48 --mem=360G --nodes=1 --time=0-1:00:00 run-toga.sh

```

#### Error: Process nextflow died
The full error code is here:

```{r died-error, eval=FALSE}
Command executed:

  sbatch .command.run

Command exit status:
  0

Command output:
  sbatch: error: (from job_submit) either --ntasks, or --ntasks-per-node, or --nodes should be provided.
  sbatch: error: Batch job submission failed: Unspecified error
  

Work dir:
  /scratch/group/songlab/emily/toga/nextflow_logs/test_chain_feats_at_1718311976/work/16/a70dc9c3f7b04b94a742de26c47b86

Tip: you can try to figure out what`s wrong by changing to the process work dir and showing the script file named `.command.sh`



Error! Process nextflow /sw/eb/sw/TOGA/2021.12.15-foss-2019b-Python-3.7.4/execute_joblist.nf --joblist /scratch/group/songlab/emily/toga/test/temp/chain_class_jobs_combined -c /scratch/group/songlab/emily/toga/nextflow_config_files/extract_chain_features_config.nf died
Program finished with exit code 1

Making gene_id: bed_data dict...
```

This error occurred when I was trying to run TOGA wihtout submitting it as a slurm job. Follow the instructions to create a file called *run-toga.sh* and submit as a batch job.

#### Error! File /sw/eb/sw/TOGA/2024.05.26-foss-2022a/CESAR2.0/cesar not found!
 This is an issue that occurs when TOGA is incorrectly installed on a cluster, contact IT to have them redo the installation.


#### Permission denied for chain files in modules
If using the --nd or --nextflow_dir flags, remove from your command.
  

