---
title: "Progress & Notes: 6/24-6/28"
site: workflowr::wflow_site
output:
  workflowr::wflow_html:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: default
    css: "theme3.css"
editor_options:
  chunk_output_type: console
---

#### TOGA

**Monday:** It seems like the error where TOGA produces no data from the given data may be due to the fact that it cannot find numpy to run analysis. I tried to run the command for *requirements.txt* as specified on the Hiller Lab GitHub, but it produced the following error:


![](assets/error-no-numpy.png)

I have emailed Michael about this error and am waiting for a response.

**UPDATE:** Michael answered and asked me to add these lines to the file nextflow_config_files/call_cesar_config_template.nf

```{r cesar-add-lines, eval=FALSE}

 process.pollInterval = '3 min'
    process.queueStatInterval = '6 min'
    process.submitRateLimit = '2/1min'

    process.clusterOptions = '--ntasks=1 --cpus-per-task=1'


```

This did not work and resulted in the same problem as before. Michael also said that numpy should be available by loading the TOGA module so that may not be the source of error. He asked me to run the test command for TOGA and that did not work. Waiting on a response.

#### make_lastz_chains

Maeva asked me to try running a script on the genomes to replace the periods with two underscores before putting them through the pipeline to see if that may fix the spacing error we're getting. The script was as follows:

```{r remove-periods, eval=FALSE}
#!/bin/bash

# Define the input and output files
input_file="input.fna"
output_file="input_np.fna"
# np = no periods

# Use grep to handle lines that are not sequence headers and sed to replace periods
grep -v "^>" "$input_file" | sed 's/\./__/g' > sequences.tmp

# Use grep to handle sequence headers
grep "^>" "$input_file" > headers.tmp

# Combine the headers and the modified sequences
cat headers.tmp sequences.tmp > "$output_file"

# Clean up temporary files
rm headers.tmp sequences.tmp

echo "Replacement complete. Output written to $output_file."
```

If this does not work we will modify the script to replace all spaces with a single underscore as well. Then as a last resort we will try to remove everything but the contig name from the line.

**UPDATE:** The no spaces script did not work, trying with the following script to replace all spaces and periods: 

```{r nospacenoperiod, eval=FALSE}

#!/bin/bash

# Define the input and output files
input_file="input.fna"
output_file="output.fna"

# Use grep to handle lines that are not sequence headers and sed to replace periods and spaces
grep -v "^>" "$input_file" | sed 's/\./__/g' | sed 's/ /_/g' > sequences.tmp

# Use grep to handle sequence headers
grep "^>" "$input_file" > headers.tmp

# Combine the headers and the modified sequences
cat headers.tmp sequences.tmp > "$output_file"

# Clean up temporary files
rm headers.tmp sequences.tmp

echo "Replacement complete. Output written to $output_file."


```

**UPDATE 2:** Removing the spaces didn't work either, using this script to remove everything from the header lines except the contig name. I responded to my comment on the Hiller Lab GitHub page and closed the issue, and Dr. Hiller responded and said that this is the easiest solution to this problem.

```{r ony-contig-ame, eval=FALSE}

#!/bin/bash

# Define the input and output files
input_file="input.fna"
output_file="output.fna"

# Use sed to process both headers and sequences in one pass
sed -e 's/^\(>\S*\).*/\1/' "$input_file" > "$output_file"

echo "Processing complete. Output written to $output_file."


```

This works but I am now running into the problem that make_lastz_chains is not allowed to run as many jobs as it wants to at a time and this causes the pipeline to crash. Emailed Michael, waiting on an answer.


**FINAL UPDATE:** It works! Now running alignments for most of the species on Maeva's table on her markdown site.


#### Barkdull Pipeline
**Tuesday:** I am currently running the second to last step of the Barkdull pipeline for OrthoFinder and I expect to have a completely finished run by tomorrow. I did have an error where it could not find BioPython, but I just added a module purge line to my .sh script and that corrected the issue. I have updated my scipt outline in my instructions to include this line for future users.

**Wednesday:** I've finished running this except for the step that cleans up the table after it's been created. I did have to rerun orthofinder because I accidentally overwrote the Orthogroups.txt file and I couldn't figure out how to get it back. The only reason I haven't run it yet is because I'm not sure where to run it or what command to use, but I just ran out of time yesterday before I was able to figure it out.

**Thursday:** Maeva said I could skip the very last step, so I'm finished with this as well, the only piece of this that isn't working is TOGA.


#### OrthoFinder Synteny Analysis 
Possible pipelines:  

```{r program-specs, echo=FALSE, message=FALSE, warning =FALSE}
library(knitr)
library(kableExtra)

# Data for the table
data <- data.frame(
  Program_name = c("GENESPACE", "Synima", "OrthoRefine", "pSONIC"),
  Last_update = c("May 2024", "January 2024", "April 2024", "2021"),
  OrthoFinder_Comp = c("Integrated", "Integrated", "Can be run both ways", "Takes OrthoFinder output"),
  Main_laguage =c("R", "Perl, Python, & R", "C++", "Python"),
  On_Cluster = c("N", "N", "N", "N"),
  GitHub_Link = c("https://github.com/jtlovell/GENESPACE?tab=readme-ov-file", "https://github.com/rhysf/Synima", "https://github.com/jl02142/OrthoRefine", "https://github.com/conJUSTover/pSONIC"),
  Notes___________________ = c("Seems to have the most doucmentation","Has a lot of documentation but runs on a lot of different languages","Likely faster since it's built on C++ but integrated plotting is simpler than Python and R", "Lab is very unresponsive, all updates are old, and not much documentation")
)

# Species info table
kable(data, "html", escape = FALSE, caption = "") %>%
  
  kable_styling() %>%
  column_spec(1, width = "30em") %>%
  column_spec(2, width = "30em") %>%
  column_spec(3, width = "30em") %>%
  column_spec(4, width = "30em") %>%
  column_spec(5, width = "15em") %>%
  column_spec(6, width = "30em") %>%
  column_spec(6, width = "50em") %>%
  row_spec(0, bold = TRUE, background = "#bfa2b0")%>%
  
  scroll_box(width="1000px", height="400px")

```
